api/sip/client.py

```diff
@@ -252,127 +248,94 @@ def __init__(self, target_server, target_port, login_user_id=None,
             escaped = self.separator
         self.separator_re = re.compile(escaped + "([A-Z][A-Z])")
 
+        self.sequence_number = 0
+
         self.login_user_id = login_user_id
         self.login_password = login_password
         if login_user_id and login_password:
             # We need to log in before using this server.
-            self.logged_in = False
             self.must_log_in = True
         else:
             # We're implicitly logged in.
-            self.logged_in = True
             self.must_log_in = False
 
-        # socket_lock controls access to the socket connection to the
-        # SIP2 server.
-        #
-        # We need to use an RLock here because both connect() and
-        # make_request() require the lock, and make_request() will end
-        # up calling connect() if there's an error.
-        self.socket_lock = threading.RLock()
-
-        # The only reason connect would be false is that we're running
-        # a unit test and don't actually want to use a server.
-        if connect:
-            self.connect()
-
-    def login(self, *args, **kwargs):
-        """Log in to the SIP server."""
-        return self.make_request(
-            self.login_message, self.login_response_parser,
-            *args, **kwargs
-        )
+    def login(self, connection):
+        """Log in to the SIP server if required."""
+        if self.must_log_in:
+            response = self.make_request(
+                connection, self.login_message, self.login_response_parser,
+                self.login_user_id, self.login_password, self.location_code
+            )
+            if response['login_ok'] != '1':
+                raise IOError("Error logging in: %r" % response)
+            return response
 
     def patron_information(self, *args, **kwargs):
         """Get information about a patron, possibly also verifying their
         password.
         """
+        connection = self.connect()
+        self.login(connection)
+        information = self._patron_information(connection, *args, **kwargs)
+        self.end_session(connection, *args, **kwargs)
+        self.disconnect(connection)
+        return information
```

Comment:
I think you can get to a point where the code looks like this:

```
with self.connect() as session:
    information = self._patron_information(session, *args, **kwargs)
return information
```

`connect()` is a context manager that creates a `SIPSession` and yields it. When the `SIPSession` is created it makes a socket connection and logs in. Once execution returns to the context manager, it ends the session and disconnects the context manager.